### PR TITLE
[BEAM-9971] Do not use context classloader.

### PR DESCRIPTION
**Pull request by ibzib**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ibzib** [comment]
Run Java Spark PortableValidatesRunner Batch

**Comment by TheNeuralBit** [comment]
There's a ReflectHelpers class in core with some utility functions for loading services and finding class loaders: https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/ReflectHelpers.java#L247

Sorry if this is just noise, I don't have much context here, just thought I'd point it out in case its helpful.

**Comment by ibzib** [comment]
Run Java Spark PortableValidatesRunner Batch

**Comment by ibzib** [comment]
retest this please

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by ibzib** [comment]
> There's a ReflectHelpers class in core with some utility functions for loading services and finding class loaders:

Interesting, though it doesn't look like we need anything fancy in this case. As long as we don't use the context class loader, I think we're fine.

Looks like the tests we expect to pass are passing now. Python2_PVR_Flink flakes are tracked by BEAM-8912 (still not sure what's causing those) and Spark is failing newly added test testFlattenWithDifferentInputAndOutputCoders2 (see BEAM-10016).

R: @robertwb

**Comment by lukecwik** [comment]
> > There's a ReflectHelpers class in core with some utility functions for loading services and finding class loaders:
> 
> Interesting, though it doesn't look like we need anything fancy in this case. As long as we don't use the context class loader, I think we're fine.
> 
> Looks like the tests we expect to pass are passing now. Python2_PVR_Flink flakes are tracked by [BEAM-8912](https://issues.apache.org/jira/browse/BEAM-8912) (still not sure what's causing those) and Spark is failing newly added test testFlattenWithDifferentInputAndOutputCoders2 (see [BEAM-10016](https://issues.apache.org/jira/browse/BEAM-10016)).
> 
> R: @robertwb

We have to be consistent within our application to use the correct class loader which is why `findClassLoader()` and variants like it exist in the Beam codebase. Many Java applications that rely on class loading do this incorrectly. I would suggest using `loadServicesOrdered(Class<T> iface)`

**Comment by ibzib** [comment]
@lukecwik It looks like `loadServicesOrdered(Class<T> iface)` uses the context class loader by default, and the context classloader is what caused this bug AFAICT. https://github.com/apache/beam/blob/b1791a2b4f44cfb6df47c2c6be7478048aaf2f2d/sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/ReflectHelpers.java#L284

We can use `loadServicesOrdered(iface, classloader)` to maintain the ordering of the result, but it looks like then no special classloader handling is applied.

https://github.com/apache/beam/blob/b1791a2b4f44cfb6df47c2c6be7478048aaf2f2d/sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/ReflectHelpers.java#L223-L224

Why is `NativeTransforms.class.getClassLoader()` incorrect?

**Comment by lukecwik** [comment]
> @lukecwik It looks like `loadServicesOrdered(Class<T> iface)` uses the context class loader by default, and the context classloader is what caused this bug AFAICT.
> 
> https://github.com/apache/beam/blob/b1791a2b4f44cfb6df47c2c6be7478048aaf2f2d/sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/ReflectHelpers.java#L284
> 
> We can use `loadServicesOrdered(iface, classloader)` to maintain the ordering of the result, but it looks like then no special classloader handling is applied.
> 
> https://github.com/apache/beam/blob/b1791a2b4f44cfb6df47c2c6be7478048aaf2f2d/sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/ReflectHelpers.java#L223-L224
> 
> Why is `NativeTransforms.class.getClassLoader()` incorrect?

Understanding thread context classloaders and their uses cases is painful:
https://www.javaworld.com/article/2077344/find-a-way-out-of-the-classloader-maze.html
https://puredanger.github.io/tech.puredanger.com/2007/06/15/classloaders/
https://impalablog.blogspot.com/2008/10/using-threads-context-class-loader-in.html

Why is the thread context class loader set in the first place?

**Comment by ibzib** [comment]
> Why is the thread context class loader set in the first place?

Spark sets it to MutableURLClassLoader. https://medium.com/@minfuyang/about-java-classloader-dynamic-class-generation-in-apache-spark-case-43a6d492c78f

**Comment by lukecwik** [comment]
> > Why is the thread context class loader set in the first place?
> 
> Spark sets it to MutableURLClassLoader. https://medium.com/@minfuyang/about-java-classloader-dynamic-class-generation-in-apache-spark-case-43a6d492c78f

It looks like the purpose is to support dynamic class loading for job submission via the --jars command. This doesn't make sense for Beam in all the portable environment types since we expect the SDK harness to load those classes (and for those classes to already be part of the classpath in EMBEDDED mode). It would make more sense for the tests/portable SparkRunner to be configured correctly.

**Comment by TheNeuralBit** [comment]
Is the approach in this PR acceptable for the short-term (and for 2.22.0 release branch)?

**Comment by lukecwik** [comment]
> Is the approach in this PR acceptable for the short-term (and for 2.22.0 release branch)?

Sure as long as someone commits to the follow-up. Adding a TODO comment would be helpful.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/NativeTransforms.java [hunk 1]
```diff
@@ -46,7 +46,9 @@
    * Returns true if an only if the Runner understands this transform and can handle it directly.
    */
   public static boolean isNative(RunnerApi.PTransform pTransform) {
-    Iterator<IsNativeTransform> matchers = ServiceLoader.load(IsNativeTransform.class).iterator();
+    Iterator<IsNativeTransform> matchers =
```
Comment (ibzib) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/NativeTransforms.java:49]:
```suggestion
    // TODO(BEAM-10109) Use default (context) classloader.
    Iterator<IsNativeTransform> matchers =
```

**Comment by ibzib** [comment]
> Sure as long as someone commits to the follow-up. Adding a TODO comment would be helpful.

I added a TODO and assigned BEAM-10109 to myself.

**Comment by ibzib** [comment]
An LGTM from either Luke or Robert should be fine.

**Pull request merged**